File: Leetcode/881. Boats to Save People/boats_to_save_people.cpp
// Leetcode 881. Boats to Save People
// https://leetcode.com/problems/boats-to-save-people/
// time complexity: O(nlogn)
// space complexity: O(n) because of sorting

// using two pointers
class Solution 
{
public:
    int numRescueBoats(vector<int>& people, int limit) 
    {
        int l = 0;
        int r = people.size() - 1;
        int result = 0;
        sort(people.begin(), people.end());
        
        while (l < r)
        {
            if (people[l] + people[r] <= limit)
            {
                l++;
            }
            
            r--;
            result++;
        }
        
        if (l == r)
        {
            result++;
        }
        
        return result;
    }
};

File: Leetcode/881. Boats to Save People/boats_to_save_people.rb
# Leetcode 881. Boats to Save People
# https://leetcode.com/problems/boats-to-save-people/

# @param {Integer[]} people
# @param {Integer} limit
# @return {Integer}
def num_rescue_boats(people, limit)
    i = 0
    j = people.length - 1
    boats = 0
    people = people.sort
    while i <= j
        if people[j] + people[i] <= limit
            j -= 1
            i += 1
        else
            j -= 1
        end
        boats += 1
    end
    
    return boats
end
